File: python/430_扁平化多级双向链表_medium.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        def child2brother(parent):
            child, nextNode = parent.child, parent.next
            parent.next, child.prev = child, parent
            parent.child = None
            while child:
                if child.child:
                    child2brother(child)
                else:
                    if child.next:
                        child = child.next
                    else:
                        break
            child.next = nextNode
            if nextNode:
                nextNode.prev = child
        
        ptr = head
        while ptr:
            if ptr.child:
                child2brother(ptr)
            else:
                ptr = ptr.next
        return head

File: cpp/430_扁平化多级双向链表_medium.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        if (!head) return head;
        Node *ptr = head;
        while (ptr != nullptr) {
            if (ptr->child != nullptr) {
                child2brother(ptr);
            }
            else {
                ptr = ptr->next;
            }
        }
        return head;
    }

    Node *child2brother(Node *parent) {
        Node *nextNode = parent->next;
        Node *child = parent->child;
        parent->child = nullptr;
        parent->next = child;
        child->prev = parent;
        while (child != nullptr) {
            if (child->child != nullptr) {
                child2brother(child);
            }
            else {
                if (child->next != nullptr)
                    child = child->next;
                else 
                    break;
            }
        }
        child->next = nextNode;
        if (nextNode != nullptr)
            nextNode->prev = child;
        return child;
    }
        
};